checker, generics: no checker error when passing incompatible type to contains

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

It should error "error: cannot use `[]string` as `string` in argument 1 to `.contains()`"

**V doctor:**

```
V full version: V 0.4.9 5b44b67
OS: macos, macOS, 15.2, 24C101
Processor: 8 cpus, 64bit, little endian, Apple M2

getwd: /Users/leodev/p/v/cluedo
vexe: /Users/leodev/p/v/v/v
vexe mtime: 2024-12-27 12:59:18

vroot: OK, value: /Users/leodev/p/v/v
VMODULES: OK, value: /Users/leodev/.vmodules
VTMP: OK, value: /tmp/v_501

Git version: git version 2.39.5 (Apple Git-154)
Git vroot status: weekly.2024.52-6-g5b44b672-dirty
.git/config present: true

CC version: Apple clang version 16.0.0 (clang-1600.0.26.6)
emcc version: N/A
thirdparty/tcc status: thirdparty-macos-arm64 713692d4
```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg repro.v`

```v

fn generic_fn[T]() {
	arr := []T{}
	println(arr.contains(arr))
}

fn main() {
	generic_fn[string]()
}
```

**What did you expect to see?**

Checker error

**What did you see instead?**

```
================== C compilation error (from cc): ==============
cc: /tmp/v_501/repro.01JG44NBAG651C6NHC362H523Z.tmp.c:5589:37: error: passing 'Array_string' (aka 'struct array') to parameter of incompatible type 'string' (aka 'struct string')
cc:  5589 |         println(Array_string_contains(arr, arr) ? _SLIT("true") : _SLIT("false"));
cc:       |                                            ^~~
cc: /tmp/v_501/repro.01JG44NBAG651C6NHC362H523Z.tmp.c:1622:58: note: passing argument to parameter 'v' here
cc:  1622 | static bool Array_string_contains(Array_string a, string v) {
cc:       |                                                          ^
cc: 1 error generated.
================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzZlYWFmNmYxNDRmNTg1YWU1MTE2ZDciLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.72Byzcem0hpzHGdRLX8K2SV_dVk4ToiWSu3MEuiiYIw">Huly&reg;: <b>V_0.6-21715</b></a></sub>

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -3568,7 +3568,7 @@ fn (mut c Checker) array_builtin_method_call(mut node ast.CallExpr, left_type as
 		if node.args.len != 1 {
 			c.error('`.contains()` expected 1 argument, but got ${node.args.len}', node.pos)
 		} else if !left_sym.has_method('contains') {
-			arg_typ := c.expr(mut node.args[0].expr)
+			arg_typ := c.unwrap_generic(c.expr(mut node.args[0].expr))
 			c.check_expected_call_arg(arg_typ, c.unwrap_generic(elem_typ), node.language,
 				node.args[0]) or {
 				c.error('${err.msg()} in argument 1 to `.contains()`', node.args[0].pos)
@@ -3582,8 +3582,9 @@ fn (mut c Checker) array_builtin_method_call(mut node ast.CallExpr, left_type as
 		if node.args.len != 1 {
 			c.error('`.index()` expected 1 argument, but got ${node.args.len}', node.pos)
 		} else if !left_sym.has_method('index') {
-			arg_typ := c.expr(mut node.args[0].expr)
-			c.check_expected_call_arg(arg_typ, elem_typ, node.language, node.args[0]) or {
+			arg_typ := c.unwrap_generic(c.expr(mut node.args[0].expr))
+			c.check_expected_call_arg(arg_typ, c.unwrap_generic(elem_typ), node.language,
+				node.args[0]) or {
 				c.error('${err.msg()} in argument 1 to `.index()`', node.args[0].pos)
 			}
 		}
@@ -3625,7 +3626,7 @@ fn (mut c Checker) array_builtin_method_call(mut node ast.CallExpr, left_type as
 		if node.args.len != 1 {
 			c.error('`.delete()` expected 1 argument, but got ${node.args.len}', node.pos)
 		} else {
-			arg_typ := c.expr(mut node.args[0].expr)
+			arg_typ := c.unwrap_generic(c.expr(mut node.args[0].expr))
 			c.check_expected_call_arg(arg_typ, ast.int_type, node.language, node.args[0]) or {
 				c.error('${err.msg()} in argument 1 to `.delete()`', node.args[0].pos)
 			}

--- a/vlib/v/checker/tests/array_generic_methods_err.out
+++ b/vlib/v/checker/tests/array_generic_methods_err.out
@@ -1,0 +1,21 @@
+vlib/v/checker/tests/array_generic_methods_err.vv:4:23: error: cannot use `[]string` as `string` in argument 1 to `.contains()`
+    2 | fn generic_fn[T]() {
+    3 |     mut arr := []T{}
+    4 |     println(arr.contains(arr))
+      |                          ~~~
+    5 |     arr.delete(arr)
+    6 |     arr.index(arr)
+vlib/v/checker/tests/array_generic_methods_err.vv:5:13: error: cannot use `[]string` as `int` in argument 1 to `.delete()`
+    3 |     mut arr := []T{}
+    4 |     println(arr.contains(arr))
+    5 |     arr.delete(arr)
+      |                ~~~
+    6 |     arr.index(arr)
+    7 | }
+vlib/v/checker/tests/array_generic_methods_err.vv:6:12: error: cannot use `[]string` as `string` in argument 1 to `.index()`
+    4 |     println(arr.contains(arr))
+    5 |     arr.delete(arr)
+    6 |     arr.index(arr)
+      |               ~~~
+    7 | }
+    8 |

--- a/vlib/v/checker/tests/array_generic_methods_err.out
+++ b/vlib/v/checker/tests/array_generic_methods_err.out
@@ -1,21 +1,21 @@
-vlib/v/checker/tests/array_generic_methods_err.vv:4:23: error: cannot use `[]string` as `string` in argument 1 to `.contains()`
-    2 | fn generic_fn[T]() {
-    3 |     mut arr := []T{}
-    4 |     println(arr.contains(arr))
+vlib/v/checker/tests/array_generic_methods_err.vv:3:23: error: cannot use `[]string` as `string` in argument 1 to `.contains()`
+    1 | fn generic_fn[T]() {
+    2 |     mut arr := []T{}
+    3 |     println(arr.contains(arr))
       |                          ~~~
-    5 |     arr.delete(arr)
-    6 |     arr.index(arr)
-vlib/v/checker/tests/array_generic_methods_err.vv:5:13: error: cannot use `[]string` as `int` in argument 1 to `.delete()`
-    3 |     mut arr := []T{}
-    4 |     println(arr.contains(arr))
-    5 |     arr.delete(arr)
+    4 |     arr.delete(arr)
+    5 |     arr.index(arr)
+vlib/v/checker/tests/array_generic_methods_err.vv:4:13: error: cannot use `[]string` as `int` in argument 1 to `.delete()`
+    2 |     mut arr := []T{}
+    3 |     println(arr.contains(arr))
+    4 |     arr.delete(arr)
       |                ~~~
-    6 |     arr.index(arr)
-    7 | }
-vlib/v/checker/tests/array_generic_methods_err.vv:6:12: error: cannot use `[]string` as `string` in argument 1 to `.index()`
-    4 |     println(arr.contains(arr))
-    5 |     arr.delete(arr)
-    6 |     arr.index(arr)
+    5 |     arr.index(arr)
+    6 | }
+vlib/v/checker/tests/array_generic_methods_err.vv:5:12: error: cannot use `[]string` as `string` in argument 1 to `.index()`
+    3 |     println(arr.contains(arr))
+    4 |     arr.delete(arr)
+    5 |     arr.index(arr)
       |               ~~~
-    7 | }
-    8 |
+    6 | }
+    7 |

--- a/vlib/v/checker/tests/array_generic_methods_err.vv
+++ b/vlib/v/checker/tests/array_generic_methods_err.vv
@@ -1,0 +1,10 @@
+fn generic_fn[T]() {
+	mut arr := []T{}
+	println(arr.contains(arr))
+	arr.delete(arr)
+	arr.index(arr)
+}
+
+fn main() {
+	generic_fn[string]()
+}

